### PR TITLE
Fix Comma Parsing Between Enumerators

### LINKED ISSUE #2514: Slice Allows Errorneous Commas in Between Enumerators
The compiler is perfectly happy to let this through:
```
module Whatever {
    enum Foo {
        Red,,,, Orange, , Blue
    }
}
```
Obviously, we should allow at-most 1 comma between 2 enumerators...

### PR DESCRIPTION
**Pull request by InsertCreativityHere**
This PR fixes #2514, and also simplifies enumerator construction in the parser.

We currently have 2 separate constructors for them, one which takes an explicit `value` parameter and one that doesn't.
I merged them into a single constructor which takes an optional now.

----

As always, feel free to ignore the generated `Grammar.cpp` file.

### CODE REVIEW

File: cpp/src/Slice/Parser.cpp [hunk 1]
```diff
@@ -2520,7 +2504,7 @@ Slice::Container::validateConstant(
     return true;
 }
 
-EnumeratorPtr
```
Comment (InsertCreativityHere) [inline cpp/src/Slice/Parser.cpp:2523]: This function _always_ returned a `nullptr`... no idea why. So I removed it.

File: cpp/src/Slice/Grammar.y [hunk 2]
```diff
@@ -1706,11 +1710,6 @@ enumerator
     auto ens = make_shared<EnumeratorListTok>(); // Dummy
     $$ = ens;
 }
-| %empty
-{
-    auto ens = make_shared<EnumeratorListTok>();
-    $$ = ens; // Dummy
-}
```
Comment (InsertCreativityHere) [inline cpp/src/Slice/Grammar.y:1713]: The culprit for the comma parsing bug.
It allowed 'empty' enumerators to appear between commas: `Red,,,, White, Blue`.

Now we allow the entire `enumerator_list` to be empty.
Much more sensible.

File: cpp/src/Slice/Parser.cpp [hunk 3]
```diff
@@ -4242,20 +4225,12 @@ Slice::Enumerator::value() const
...
-{
-}
-
-Slice::Enumerator::Enumerator(const ContainerPtr& container, const string& name, int value)
+Slice::Enumerator::Enumerator(const ContainerPtr& container, const string& name, optional<int> value)
     : SyntaxTreeBase(container->unit()),
       Contained(container, name),
-      _explicitValue(true),
-      _value(value)
+      _explicitValue(value)
```
Comment (bernardnormier) [inline cpp/src/Slice/Parser.cpp:4231]: I would find _hasExplicitValue much clearer.

Here, I have the impression you store value in _explicitValue.

Comment (bernardnormier) [reply cpp/src/Slice/Parser.cpp:4231]: Or maybe we should fix the API (in a follow-up PR) with simply:

```cpp
optional<int> Slice::Enumerator::value() const { return _value; }
```

where value is an `optional<int>`.

**Review by bernardnormier — APPROVED** [review]

File: cpp/src/Slice/Parser.cpp (hunk 3 above)
Comment (InsertCreativityHere) [reply cpp/src/Slice/Parser.cpp:4231]: I renamed it to `_hasExplicitValue`!

But your proposed API isn't quite correct.
Every enumerator _must_ have a value, the code I changed is for how the parser signals:
"hey this enumerator needs an _implicit_ value calculated for it."

In a separate part of the code we check if `_value == -1`, and if so, assign it a value of `last enumerator's value + 1`

**Pull request merged**